crates/namada/src/ledger/protocol/mod.rs

```diff
@@ -217,19 +237,45 @@ where
                 wrapper_args,
             )
             .map_err(|e| Error::WrapperRunnerError(e.to_string()))?;
-            let mut inner_res = apply_wasm_tx(
-                tx,
-                &tx_index,
-                ShellParams {
-                    tx_gas_meter,
-                    state,
-                    vp_wasm_cache,
-                    tx_wasm_cache,
-                },
-            )?;
 
-            inner_res.wrapper_changed_keys = changed_keys;
-            Ok(inner_res)
+            // Replay protection check on the batch
+            let tx_hash = tx.raw_header_hash();
+            if state.write_log().has_replay_protection_entry(&tx_hash) {
+                // If the same batch has already been committed in
+                // this block, skip execution and return
+                return Err(Error::ReplayAttempt(tx_hash));
+            }
+
+            for cmt in tx.commitments() {
+                match apply_wasm_tx(
+                    tx.batch_ref_tx(cmt),
+                    &tx_index,
+                    ShellParams {
+                        tx_gas_meter,
+                        state,
+                        vp_wasm_cache,
+                        tx_wasm_cache,
+                    },
+                ) {
+                    Err(e @ Error::GasError(_)) => {
+                        // Gas error aborts the execution of the entire batch
+                        state.write_log_mut().drop_tx();
+                        return Err(e);
+                    }
+                    res => {
```

Comment:
Currently we continue in case of other errors even for atomic batches which is not needed since the batch will fail anyway. The reason here was to mimic what we do when evaluating VPs, which is to continue until we have gas at our disposable even if one of the VPs has already rejected the tx

Reply:
I missed this comment in relation to the atomic batches.

> which is to continue until we have gas at our disposable even if one of the VPs has already rejected the tx

sorry I didn't get this

Reply:
Yes I'm not sure about this logic. I'd like to hear other opinions on this, we could also decide to short circuit the evaluation in this case

Reply:
Updated the logic in 8ab0f38e6596fdf3da3d2492c866c992fb7f18b4